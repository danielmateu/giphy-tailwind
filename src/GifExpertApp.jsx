import React, { useState } from 'react'
import { AddCategory } from './components/AddCategory'
import { GifGrid } from './components/GifGrid'

export const GifExpertApp = () => {

    const [categories, setCategories] = useState(['One Punch'])

    // AÃ±adir categorias
    const handleAdd = (newCategory) => {
        //Pasar new category a lowercase
        // newCategory = newCategory.toLowerCase()
        //Si newCategory ya existe en el arreglo
        if (categories.includes(newCategory)) return
        setCategories(cats => [newCategory, ...cats])
        
    }


    return (
        <div className='bg-sky-400 p-10'>
            {/* TITULO */}
            <h1 className='text-4xl text-center'>GifExpertApp</h1>
            {/* INPUT */}
            <AddCategory
                onNewCategory={handleAdd}
            // setCategories={setCategories}
            // categories = {categories}
            />

            {/* LISTA DE GIFS */}
            <h2 className='p-2'>Listado</h2>

            {
                categories.map(category =>
                (
                    <GifGrid
                        key={category}
                        category={category}
                    />
                )
                )
            }



            {/* GIF ITEM */}
        </div>
    )
}
