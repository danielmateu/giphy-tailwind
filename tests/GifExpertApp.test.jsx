import { render, screen, fireEvent } from "@testing-library/react";
import { GifExpertApp } from "../src/GifExpertApp";

//Pruebas en el componente GifExpertApp
describe('Pruebas en el componente GifExpertApp', () => {

    const category = 'One Punch'

    //setup function: para no repetir el setup de las pruebas repetidas veces
    const setup = () => {
        render(<GifExpertApp />);
        const input = screen.getByRole("textbox");
        const form = screen.getByRole("form", { name: "form" });
        return {
            input,
            form,
        };
    }

    //Debe hacer match con el snapshot
    test('Debe hacer match con el snapshot', () => {
        const { container } = render(<GifExpertApp />)
        expect(container).toMatchSnapshot();
    })

    //Debe agregar una categoria
    test('Debe agregar una categoria', () => {
        const { input, form } = setup()

        fireEvent.change(input, { target: { value: category } })
        fireEvent.submit(form)
        expect(screen.getByText(category)).toBeTruthy()
    })

    //Debe mostrar una lista de categorias
    test('Debe mostrar una lista de categorias', () => {
        const { input, form } = setup()

        fireEvent.change(input, { target: { value: category } })
        fireEvent.submit(form)
        expect(screen.getByText(category)).toBeTruthy()
    })

    //Debe añadir una categoria si no está repetida
    test('Debe añadir una categoria si no está repetida', () => {
        const { input, form } = setup()

        fireEvent.change(input, { target: { value: category } })
        fireEvent.submit(form)
        fireEvent.change(input, { target: { value: category } })
        fireEvent.submit(form)
        expect(screen.queryAllByText(category).length).toBe(1)
    })

    //No debe añadir la categoria si está repetida
    test('No debe añadir la categoria si está repetida', () => {
        const { input, form } = setup()

        fireEvent.change(input, { target: { value: category } })
        fireEvent.submit(form)
        fireEvent.change(input, { target: { value: category } })
        fireEvent.submit(form)
        expect(screen.queryAllByText(category).length).toBe(1)
    })

})