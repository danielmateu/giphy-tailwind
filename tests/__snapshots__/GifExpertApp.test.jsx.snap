// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente GifExpertApp Debe hacer match con el snapshot 1`] = `
<div>
  <div
    class="bg-sky-200 "
  >
    <h1
      class="text-4xl text-center"
    >
      GifExpertApp
    </h1>
    <form
      aria-label="form"
    >
      <input
        class="bg-gray-200 p-2 rounded-md m-2"
        placeholder="AÃ±ade algo..."
        type="text"
        value=""
      />
    </form>
    <div
      class="py-10"
    >
      <h3
        class="text-2xl text-center pb-4"
      >
        One Punch
      </h3>
      Loading...
      <div
        class=" px-16 grid grid-cols-1 sm:grid-cols-2 md:grid-cols-3 lg:grid-cols-4 gap-2 items-start place-items-center "
      />
    </div>
  </div>
</div>
`;
